### ISSUE TITLE
typescript-best-practices/no-extra-non-null-assertion: false positives on `!!x` with optional chaining; regions span whole method bodies and duplicate

### ISSUE THREAD

**Issue by SebastianBienert** [DataDog/datadog-static-analyzer#962]
**Describe the bug**

`typescript-best-practices/no-extra-non-null-assertion` reports `Forbidden extra non-null assertion.` on expressions that contain no non-null assertion at all. In our Angular 22 / Nx 23 monorepo it produced 4 findings, all false.

The rule is not broken in general — it correctly flags a genuine case (verified: `return v!!.a!.b;` → 1 finding). The false positives all sit on expressions that combine the `!!value` boolean-coercion idiom with optional chaining. Identifiers below are renamed, structure is unchanged:

```ts
return !!versions && versions?.length > 1;                                  // flagged
return !!excludedScenarios?.length;                                         // inside a flagged span
const hasEvents = this.getViewConfig(x?.type)?.findIndex((a) => a.title === CAT) !== -1;  // start of a flagged span
```

Two further problems with the reports themselves, independent of the matching question:

1. **The regions span whole method bodies.** In one 878-line file the reported SARIF regions are `484:93 → 510:14` and `695:23 → 709:21`. The first starts at the `{` of a method signature and ends 27 lines later at the closing brace; the second covers 15 lines and ends inside a later statement. A non-null assertion is a two-character postfix operator, so a region covering an entire body cannot be pointing at one.
2. **Identical findings are emitted twice.** The `484:93 → 510:14` finding appears twice in the same SARIF run, with the same rule, message and region.

**Probable cause**

The rule, as served by `/api/v2/static-analysis/rulesets/typescript-best-practices`:

```
((non_null_expression (non_null_expression)+) @expression)
((non_null_expression (parenthesized_expression (non_null_expression)+)) @expression)
(((non_null_expression) (optional_chain)) @expression)
(((parenthesized_expression (non_null_expression)) (optional_chain)) @expression)
(((non_null_expression) "?.") @expression)
((((parenthesized_expression (non_null_expression))) "?.") @expression)
```

```js
function report(node, message) {
  addError(buildError(node.start.line, node.start.col, node.end.line, node.end.col, message));
}

function visit(query, filename, code) {
  const node = query.captures["expression"];
  if (node) {
    report(node, 'Forbidden extra non-null assertion.');
  }
}
```

The last four alternatives are **sibling sequences with no parent anchor** — `((A) (B))` matches an `A` followed by a `B` under any parent, including sibling *statements*. `buildError` then spans from the first sibling's start to the last sibling's end, which would explain regions covering a whole method body, and the duplicate report is consistent with two alternatives matching the same pair.

The `visit` body also reports unconditionally on any capture, with no check that what matched is actually a doubled assertion (`x!!`) or an assertion followed by `?.` on the *same* member chain.

**To Reproduce**

I could not isolate a small reproducer, and I want to be straight about that rather than post something that does not reproduce:

- The flagged expressions in isolation produce **0 findings** (5 variants tried, e.g. `export function f(v?: string[]): boolean { return !!v && v?.length > 1; }`).
- A validity-constrained delta-debugging pass (every candidate required to parse with 0 diagnostics via `ts.createSourceFile`, and to still report the finding) stalled at 582 lines, which still contains proprietary imports and so is not publishable.
- The trigger therefore appears to depend on parse context, the same way as the `no-empty-pattern` false positives in #961.

What reproduces deterministically on our side: a single 878-line file of valid TypeScript (0 TS parse diagnostics), analyzed with

```yaml
schema-version: v1.4
sast:
  use-default-rulesets: false
  use-rulesets:
    - typescript-best-practices
```

via `datadog-static-analyzer -i . -o out.sarif -f sarif`, yields the 3 regions above (4 findings including the duplicate). I'm happy to work with a maintainer on sharing a reproducer, or to test a patched ruleset against the real file and report back.

**Expected behavior**

- No finding for `!!value && value?.length > 1` — `!!` is boolean coercion, not a duplicated non-null assertion.
- When a finding is legitimate, the region should cover the offending assertion, not an enclosing method body.
- The same violation should not be reported twice for one location.

Cross-check: `@typescript-eslint/no-extra-non-null-assertion` is enabled at error severity in the same repo (`eslint --print-config` shows `[2]`) and reports zero problems on every file this rule flags.

**Suggested fix**

Anchor the sibling alternatives to a parent node (for example `member_expression` / `subscript_expression` / `call_expression`), so that a `non_null_expression` followed by `?.` only matches within one member chain rather than across arbitrary siblings, and validate in `visit` that the capture really is a doubled assertion before reporting.

**Additional context**

Same ruleset, same shape of problem as #961 (`no-empty-pattern` matching by node type without positional validation). Filed separately since it is a different rule with a different query.

**Environment**

- `datadog-static-analyzer` 0.9.1, `x86_64-pc-windows-msvc` (installed via mise, `github:DataDog/datadog-static-analyzer`)
- OS: Windows 11
- `DD_SITE=datadoghq.eu`; ruleset `typescript-best-practices`, rule `no-extra-non-null-assertion`
- Source project: Angular 22 / Nx 23 monorepo, TypeScript 6.0.3

Workaround in use: the rule is disabled repo-wide via `ruleset-configs` → `rule-configs` → `ignore-paths: ["**"]`, since the false positives track parse context rather than specific source lines.

**Comment by jasonforal** [comment]
(As mentioned in #961, we'll be able to take a look at this next week)
